numpy/core/src/multiarray/common.c

```diff
@@ -946,4 +947,344 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
+
+    /* Allocate interned string storage and NULL it for error handling */
+    PyObject **kw_strings = malloc(sizeof(PyObject *) * (nkwargs + 1));
+    if (kw_strings == NULL) {
+        PyErr_NoMemory();
+        return -1;
+    }
+    memset(cache->kw_strings, 0, sizeof(PyObject *) * (nkwargs + 1));
+
+    va_copy(va, va_orig);
+    for (int i = 0; i < nargs; i++) {
+        /* Advance through non-kwargs, which do not require setup. */
+        char *name = va_arg(va, char *);
+        convert *converter = va_arg(va, convert *);
+        va_arg(va, void *);
+
+        /*
+        if (converter == NULL) {
+            cache->intent[i] = 'O';
+        }
+        else if (converter == (convert *)PyArray_BoolConverter) {
+            cache->intent[i] = '?';
+        }
+        else if (converter == (convert *)PyArray_PythonPyIntFromInt) {
+            cache->intent[i] = 'i';
+        }
+        else {
+            cache->intent[i] = '&';
+        }
+         */
+
+        if (i >= npositional_only) {
+            int i_kwarg = i - npositional_only;
+            cache->kw_strings[i_kwarg] = PyUString_InternFromString(name);
+            if (cache->kw_strings[i_kwarg] == NULL) {
+                va_end(va);
+                goto error;
+            }
+        }
+    }
+    va_end(va);
+    return 0;
+
+error:
+    for (int i = 0; i < nkwargs; i++) {
+        Py_XDECREF(cache->kw_strings[i]);
+    }
+    return -1;
+}
+
+static int
+raise_incorrect_number_of_positional_args(const char *funcname,
+        const _NpyArgParserCache *cache, Py_ssize_t len_args)
+{
+    // TODO: The error message will also note the number of keyword args!
+    if (cache->npositional == cache->nrequired) {
+        PyErr_Format(PyExc_TypeError,
+                "%s() takes %zd positional arguments but %zd were given",
+                funcname, cache->npositional, len_args);
+    }
+    else {
+        PyErr_Format(PyExc_TypeError,
+                "%s() takes %zd to %zd positional arguments but %zd "
+                "were given",
+                funcname, cache->nrequired, cache->npositional, len_args);
+    }
+    return 0;
+}
+
+static void
+raise_missing_argument(const char *funcname,
+        const _NpyArgParserCache *cache, int i)
+{
+    if (i < cache->npositional_only) {
+        PyErr_Format(PyExc_TypeError,
+                "%s() missing required positional argument %d",
+                funcname, i);
+    }
+    else {
+        PyObject *kw = cache->kw_strings[i-cache->npositional_only];
+        PyErr_Format(PyExc_TypeError,
+                "%s() missing required argument '%S' (pos %d)",
+                funcname, kw, i);
+    }
+}
+
+/*
+ * Find key in a list of pointers to keyword names.
+ * The list should end with NULL.
+ *
+ * Returns either the index into the list (pointing to the final key with NULL
+ * if no match was found), or -1 on failure.
+ */
+static NPY_INLINE npy_intp
+locate_key(PyObject *const *kwnames, PyObject *key)
```

Comment:
Hmm, this was the ufunc code, my code was slightly different... A bit of a mess up here, so it might change (although making it into an inlined function may be nicer.